Roles and Permissions

etiennemarais opened the pull request:
## Summary

We wanted to have a basic set of user roles and permissions to allow for varying behaviour of the system based on a user's assigned role and/or permissions.

## Story map and permission's list

**Use cases**
![Screenshot 2020-04-15 at 20 13 19](https://user-images.githubusercontent.com/4479918/79372024-9c553900-7f55-11ea-8fa2-e465f160a557.png)

**Roles/Permissions list**
![Screenshot 2020-04-15 at 20 13 25](https://user-images.githubusercontent.com/4479918/79372027-9fe8c000-7f55-11ea-9ae2-b25b87f2b2f0.png)


## Changes

- Added a migration to add an organiser identifier on a user to aid in limiting showing of other events.
- Fixed the user invite from the account screen to add the invited user to be added to the same organiser as the person inviting them.
- Added baseline roles/permissions package
- Upgraded laravel to latest stable 6.x
- Fixed tests that broke during the change over
- Added new landing view for normal users to see a list of their events
- Tweaked the UI to show/hide certain organiser and account level options based on the user's permissions or role
- Added a base test seed to add the new roles/permissions and assign them to the roles they belong
- Cleaned up the error pages to use the laravel simple styles
- Added an initial seed for roles/permissions to the migrations
- Scan the db for legacy users and assign the default `user` role to them, `super admin` needs to be assigned manually
- Added UI to invite a new user with their organiser and role selected
- Added UI to show the role for a user and toggle between them
- Added UI to toggle between adding the extended `can manage events` permission to users based on clicking the checkbox


### Demo

If you want to test and check out the branch you can run these commands to get you started on a fresh database:

```sh
composer install
php artisan migrate:refresh --seed && php artisan db:seed --class=RolesPermissionsLocal
```

That will demo seed a bunch of users to try out with their roles already set (we don’t have a dedicated UI for them yet)

```sh
u: super@test.com
p: pass
This user can do everything

u: user1@test.com
p: pass
This user can manage all events within an organiser level

u: user2@test.com
p: pass
This user can only see events created by them even though they belong to the same organiser 

u: checkin@test.com
p: pass
This user can see all events for an organiser but are limited to their check in pages only. 
```

## Screenshots
**Can toggle the role and extended permissions**
![Screenshot 2020-06-27 at 22 28 30](https://user-images.githubusercontent.com/4479918/85931589-b8664680-b8c5-11ea-92f7-7ee29de710a5.png)

**Can select the organiser this user belongs to as well as the Role**
![Screenshot 2020-06-27 at 22 28 22](https://user-images.githubusercontent.com/4479918/85931590-ba300a00-b8c5-11ea-9960-3f65098fb76f.png)

---
**The whole new Role/Permission assignment UI - Only available to super admins**
![Screenshot 2020-06-27 at 22 28 13](https://user-images.githubusercontent.com/4479918/85931591-bb613700-b8c5-11ea-9262-11d75e315a26.png)

Leen15 committed:
Add docker production environment

Leen15 committed:
fix typos

ftbastler committed:
Setup German translation and translate all emails

Modding committed:
German Translation of Footer.blade.php

Modding committed:
Translation of Emails

Modding committed:
lang de erstellt und die ersten Dateien bearbeitet.

Modding committed:
weitere Dateien bearbeitet

AdrienPoupa committed:
Prevent duplicate queries

AdrienPoupa committed:
Update FirstRunMiddleware.php

AdrienPoupa committed:
Remove some duplicate queries

Modding committed:
Fees bearbeitet.

Modding committed:
Installer.php bearbeitet

Modding committed:
Installer.php fertiggestellt

Modding committed:
Javascript.php fertiggestellt

Modding committed:
anpassungen

Modding committed:
erste änderungen

Modding committed:
Public_ViewEvent.php bearbeitet

Modding committed:
Rechtschreibfehler

Modding committed:
anpassungen

bagage committed:
Do not escape html for post order message

publicarray committed:
Fix XSS and 500 error if version check fails

falcecamogli committed:
First commit Italian localization

falcecamogli committed:
Fixed missed translation in Italian localization

falcecamogli committed:
Added missing translations for Italian localization

falcecamogli committed:
Fixed wrong translations for Italian localization

falcecamogli committed:
Enable Italian locale

falcecamogli committed:
Added missing translatable strings in Stripe, also for other languages

falcecamogli committed:
Fix translation for Italian localization

falcecamogli committed:
Added missing translation label

samdb committed:
Merge branch 'master' into version-check-fixes

johannac committed:
Merge branch 'develop' into fix-post-order-display-message-escaping

johannac committed:
Merge pull request #629 from bagage/fix-post-order-display-message-escaping

Do not escape html for post order message

johannac committed:
Merge branch 'develop' into version-check-fixes

johannac committed:
Add installed function back to utils to fix 500 error in tests

johannac committed:
Consolidate two getVersionInfo functions into 1

johannac committed:
Add unit test for new parse_version function

johannac committed:
Fix types in new parse_version function and expand regex a bit

johannac committed:
Change version check to use github version file instead of old attendize website

johannac committed:
Merge pull request #743 from publicarray/version-check-fixes

Fix XSS and 500 error if version check fails

chrisbrosnan committed:
Update ManageAccountController.php

Corrected 500 internal server error in accessing Account Settings in top right dropdown options, caused by Validator being used when already in use within the same controller (app/Http/Controllers/ManageAccouuntController.php). No problems seem to be caused by this fix.

johannac committed:
Merge pull request #789 from chrisbrosnan/develop

Update ManageAccountController.php

johannac committed:
Fix installer so it shows default values in the form

johannac committed:
Refactor docker config so nginx and php run on 1 container.
Add targets to the dockerfile so we can build multiple images, for web and for worker, with the same base image.
Add makefile for running commands more easily.

johannac committed:
Add some comments to the docker set up

johannac committed:
Add apache support

Commit:
Update dependencies and install roles and permissions package

Commit:
Add better minimal error pages

Commit:
Add permissions config, Add roles and permissions migrations. Add organiser identifier to user table migration

Commit:
Link user to organiser. Fetch events with organisers

Commit:
Add missing translations. Wrap UI parts in role/permissions checks to conditionally render based on the assigned roles or permissions

Commit:
Add organiser to user in the test seeds. Add basic seeds for roles and permissions to test locally. Will make them into migrations that run next

Commit:
Wrap organiser routes in the correct middleware checks for roles

Commit:
Add CanManageOrganisers middleware. Implicitly allow super admin users to access everything. Cleanup ad hoc spacings. Fix the firstrunmiddleware to check if any organisers exist and redirect to the right place

Commit:
Cleanup the database seeder class names. Update the index controller to, based on the role, show events first instead of the organiser selection screen, if the user does not have manage organisers permissions. Update the account manage controller to assign a normal user role to a new user.

Commit:
Bump version to 2.1.0

justynpride commented:
@etiennemarais I'm very excited for this feature, you've done a great job. I appear to be getting an error when trying to invite a new user:

`[2020-05-30 15:53:59] production.ERROR: Argument 1 passed to App\Http\Controllers\ManageAccountController::App\Http\Controllers\{closure}() must be an instance of Illuminate\Contracts\Mail\Mailable, instance of Illuminate\Mail\Message given, called in /home/afamily1/public_html/attendize_test/vendor/laravel/framework/src/Illuminate/Mail/Mailer.php on line 244 {"userId":1,"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalThrowableError(code: 0): Argument 1 passed to App\\Http\\Controllers\\ManageAccountController::App\\Http\\Controllers\\{closure}() must be an instance of Illuminate\\Contracts\\Mail\\Mailable, instance of Illuminate\\Mail\\Message given, called in /home/afamily1/public_html/attendize_test/vendor/laravel/framework/src/Illuminate/Mail/Mailer.php on line 244 at /home/afamily1/public_html/attendize_test/app/Http/Controllers/ManageAccountController.php:207)`

This may be unrelated to your code. It is a clean install of the system.

Commit:
Fix regression in sending mail to a new user

etiennemarais commented:
@justynpride Thank you. I initially made the PR in a different place and this was a small regression when I ported it over. I pushed a fix for it 🙏

justynpride commented:
> @justynpride Thank you. I initially made the PR in a different place and this was a small regression when I ported it over. I pushed a fix for it 🙏
Perfect! This is a great additional to the system.

johannac commented:
@etiennemarais this looks great! Thank you! I will review and test on Monday. I am planning to tag a release on June 1st. I'm aiming to do a minor release once a month. Do you want to try and get this into the June release, or should we save it for July so we have some more time to test it?

etiennemarais commented:
> I am planning to tag a release on June 1st. I'm aiming to do a minor release once a month.

This is great news. Please let me know what help you need and I will jump in 🙏 

> Do you want to try and get this into the June release, or should we save it for July so we have some more time to test it?

I have a feeling I can do a bit better with this PR so lets bump it until next _(July)_ release train please. The reason I say that is to solve some of the unsolved issues before I can sign this off.

- At the moment there is no UI to assign roles to users OR permissions to roles and I'd like to fix that even though it's a simple design.
- I need some help deciding about what to do with legacy install `users` and which roles to assign them. (There are two possible options here for me but I'd like to hear your thoughts first please)
- Show/Hide stats based on role or if that is even relevant for now.

To be perfectly honest I think we should push to review/test and get #723 into the next release cycle @johannac in terms of priority 🙏 It adds a lot of value

johannac committed:
Fix the copy env file in makefile

johannac committed:
Version bump

johannac committed:
Remove old compose file

publicarray committed:
Add hCaptcha (#793)

* Add hcaptcha

* Add hcaptcha to contact form

* remove some duplicate code

* fix use

* add hcaptcha

Co-authored-by: Johanna Cherry <johanna@johannacherry.com>

publicarray committed:
Add reCAPTCHA (#585)

* Add optional google recaptcha on login

* use recaptcha domain and added recaptcha to signup

* fix recapture on signup

* fix Undefined property: stdClass error

* fix bool operation

* Fix optional captcha

* env -> config (recaptcha)

fix php artisan config:cache

Co-authored-by: Johanna Cherry <johanna@johannacherry.com>

SamuelNitsche committed:
Update Fees.php (#810)

melvinthoabala committed:
Allow use of both reCaptcha hCaptcha or no capture based on config (#808)

* WIP: add config for captcha and allow user to choose between them

* add captha on sign up

* clean up

* read captcha switch from env

johannac committed:
Add min reqs to the readme file

johannac committed:
Add commands to makefile for clearing laravel cache and recompiling autoloaders

johannac committed:
Add artisan migrate to the setup scripts in case there's database migrations that need to be run

johannac committed:
Last name is required by the database schema so needs to be required by the signup form too

johannac committed:
Fix redirect on the signup page - the import was missing, but also the redirect was being ignored all togeher because you can only do a redirect in a route itself, if it's in the constructor you have to explicitly call send.

johannac committed:
Change the use of Utils::isAttendize() to Utils::isCloud() during the signup process - this is used to hide the signup form in cloud environments, this way it will show in dev.

johannac committed:
Remove dupe translation

johannac committed:
Refactor a bit of the captcha logic for the signup page
- move services to subdirectory
- create recaptcha service
- create factory for captcha services
- create partial views for each captcha service and include in the main view
- standardize config options

johannac committed:
Refactor login page to use captcha services

johannac committed:
Refactor contact form page to use captcha services

scottyeung committed:
Use promises to initialise camera for check-in (#712)

* Update QRCode Reader to use navigator.mediaDevices

Update to use navigator.mediaDevices for Safari and Chrome Desktop to enable using camera

* Update public/assets/javascript/check_in.js

Co-authored-by: Etienne Marais <etiennemarais@users.noreply.github.com>

justynpride commented:
Hi @etiennemarais. I was wondering how this amazing new feature was progressing, and if it might make it into the next release, especially if there is a July one? If you need any help testing, etc.. I'm available.

Commit:
Merge branch 'develop' into feature/roles-and-permissions

Commit:
Add seeder to run with migration to pre seed default roles and permissions and try to attach the default user role to legacy users in one pass

etiennemarais commented:
@justynpride I pushed a change to attach the `user` role to any legacy users. Also testing would help please. I am still busy with the UI to assign roles to users and should be able to push that today if all goes well. The last couple of weeks have been really busy at work so sorry about the delay. 🙏

Commit:
Update the account users UI with the roles assigned to the users in the list. General cleanup of the template

Commit:
Update the invite user form to take a first/last name along with email, which organiser they can belong to and their role

Commit:
Layout the table better. Default select the user role for sensible defaults. Make a space for adding additional permissions to a user role like managing multiple events but still not being a super admin

justynpride commented:
> @justynpride I pushed a change to attach the `user` role to any legacy users. Also testing would help please. I am still busy with the UI to assign roles to users and should be able to push that today if all goes well. The last couple of weeks have been really busy at work so sorry about the delay. 🙏

@etiennemarais Am exciting to test this. I've trying it on a fresh install of master (works), and then checkout out this branch. I've rerun composer install, and then run to migrate the db. It appears to hang on this:

 Migrating: 2020_04_09_063112_create_permission_tables

Am very happy to have got something wrong, but thought it worth letting you know.

Commit:
Update the user management screen with role selection dropdowns. Bind an event to the assigned role dropdown to sync the new roles/permissions based on the selection update. Add new route to allow updating single user roles. Users are allowed to have only a single role.

etiennemarais commented:
> It appears to hang on this

Could you see if there are any errors in the log file? That migration tries to scan the db for users and assign the default role to them.

justynpride commented:
> > It appears to hang on this
> 
> Could you see if there are any errors in the log file? That migration tries to scan the db for users and assign the default role to them.
This is what is shown in the Attendize logs:

`[2020-06-27 19:05:31] production.DEBUG: select * from information_schema.tables where table_schema = ? and table_name = ? and table_type = 'BASE TABLE' ["tick_tickets","migrations"] 
[2020-06-27 19:05:31] production.DEBUG: select `migration` from `migrations` order by `batch` asc, `migration` asc  
[2020-06-27 19:05:31] production.DEBUG: select max(`batch`) as aggregate from `migrations`  
[2020-06-27 19:05:31] production.DEBUG: create table `permissions` (`id` bigint unsigned not null auto_increment primary key, `name` varchar(255) not null, `guard_name` varchar(255) not null, `created_at` timestamp null, `updated_at` timestamp null) default character set utf8mb4 collate 'utf8mb4_unicode_ci'  
[2020-06-27 19:05:31] production.DEBUG: create table `roles` (`id` bigint unsigned not null auto_increment primary key, `name` varchar(255) not null, `guard_name` varchar(255) not null, `created_at` timestamp null, `updated_at` timestamp null) default character set utf8mb4 collate 'utf8mb4_unicode_ci'  
[2020-06-27 19:05:32] production.DEBUG: create table `model_has_permissions` (`permission_id` bigint unsigned not null, `model_type` varchar(255) not null, `model_id` bigint unsigned not null) default character set utf8mb4 collate 'utf8mb4_unicode_ci'  
[2020-06-27 19:05:32] production.DEBUG: alter table `model_has_permissions` add index `model_has_permissions_model_id_model_type_index`(`model_id`, `model_type`)  
[2020-06-27 19:05:32] production.DEBUG: alter table `model_has_permissions` add constraint `model_has_permissions_permission_id_foreign` foreign key (`permission_id`) references `permissions` (`id`) on delete cascade  
[2020-06-27 19:05:32] production.DEBUG: alter table `model_has_permissions` add primary key `model_has_permissions_permission_model_type_primary`(`permission_id`, `model_id`, `model_type`)  
[2020-06-27 19:05:32] production.DEBUG: create table `model_has_roles` (`role_id` bigint unsigned not null, `model_type` varchar(255) not null, `model_id` bigint unsigned not null) default character set utf8mb4 collate 'utf8mb4_unicode_ci'  
[2020-06-27 19:05:32] production.DEBUG: alter table `model_has_roles` add index `model_has_roles_model_id_model_type_index`(`model_id`, `model_type`)  
[2020-06-27 19:05:32] production.DEBUG: alter table `model_has_roles` add constraint `model_has_roles_role_id_foreign` foreign key (`role_id`) references `roles` (`id`) on delete cascade  
[2020-06-27 19:05:32] production.DEBUG: alter table `model_has_roles` add primary key `model_has_roles_role_model_type_primary`(`role_id`, `model_id`, `model_type`)  
[2020-06-27 19:05:32] production.DEBUG: create table `role_has_permissions` (`permission_id` bigint unsigned not null, `role_id` bigint unsigned not null) default character set utf8mb4 collate 'utf8mb4_unicode_ci'  
[2020-06-27 19:05:32] production.DEBUG: alter table `role_has_permissions` add constraint `role_has_permissions_permission_id_foreign` foreign key (`permission_id`) references `permissions` (`id`) on delete cascade  
[2020-06-27 19:05:32] production.DEBUG: alter table `role_has_permissions` add constraint `role_has_permissions_role_id_foreign` foreign key (`role_id`) references `roles` (`id`) on delete cascade  
[2020-06-27 19:05:32] production.DEBUG: alter table `role_has_permissions` add primary key `role_has_permissions_permission_id_role_id_primary`(`permission_id`, `role_id`)  
[2020-06-27 19:05:32] production.DEBUG: Seeding initial roles and permissions from the migration  
`

Terminal never shows the migration as completing. When I go to my install in the browser the error shown is :

`Symfony\Component\Debug\Exception\FatalThrowableError
Call to a member function events() on null`

`App\Http\Controllers\IndexController::showIndex
app/Http/Controllers/IndexController.php:35`

Commit:
Allow super admin users to give permission to User role types to manage events within their organiser that weren't created by them. This operation toggles between allowing and revoking the can manage events permission based ont he checkbox on the user manage screen

Commit:
Clean up unused javascript

etiennemarais commented:
@johannac, @samdb  This PR is ready for review please 🙏 

@justynpride Please re-checkout this branch, i've made some changes. I wonder if there is an exception that gets swallowed and not handled properly. Thanks for brining it up, Lets fix it together.

Have you run a `composer dump-autoload` recently? There are a couple of new Seeder classes that got added.

justynpride commented:
> @justynpride Please re-checkout this branch, i've made some changes. I wonder if there is an exception that gets swallowed and not handled properly. Thanks for brining it up, Lets fix it together.
> 
> Have you run a `composer dump-autoload` recently? There are a couple of new Seeder classes that got added.
@etiennemarais I've not run `composer dump-autoload`, but have just done so, but it does appear to have resolved it. 

What I have just done is to do a fresh install and checked out your branch before running the Attendize installer. Installed runs but gives an error message of `Doctrine\DBAL\Schema\SchemaException
There is no column with name 'recipients' on table 'messages'.`

Looking at the logs I have these records (I have truncated so you have just the last few):

`[2020-06-28 15:10:31] production.DEBUG: create table `permissions` (`id` bigint unsigned not null auto_increment primary key, `name` varchar(255) not null, `guard_name` varchar(255) not null, `created_at` timestamp null, `updated_at` timestamp null) default character set utf8mb4 collate 'utf8mb4_unicode_ci'  
[2020-06-28 15:10:31] production.DEBUG: create table `roles` (`id` bigint unsigned not null auto_increment primary key, `name` varchar(255) not null, `guard_name` varchar(255) not null, `created_at` timestamp null, `updated_at` timestamp null) default character set utf8mb4 collate 'utf8mb4_unicode_ci'  
[2020-06-28 15:10:31] production.DEBUG: create table `model_has_permissions` (`permission_id` bigint unsigned not null, `model_type` varchar(255) not null, `model_id` bigint unsigned not null) default character set utf8mb4 collate 'utf8mb4_unicode_ci'  
[2020-06-28 15:10:31] production.DEBUG: alter table `model_has_permissions` add index `model_has_permissions_model_id_model_type_index`(`model_id`, `model_type`)  
[2020-06-28 15:10:31] production.DEBUG: alter table `model_has_permissions` add constraint `model_has_permissions_permission_id_foreign` foreign key (`permission_id`) references `permissions` (`id`) on delete cascade  
[2020-06-28 15:10:31] production.DEBUG: alter table `model_has_permissions` add primary key `model_has_permissions_permission_model_type_primary`(`permission_id`, `model_id`, `model_type`)  
[2020-06-28 15:10:31] production.DEBUG: create table `model_has_roles` (`role_id` bigint unsigned not null, `model_type` varchar(255) not null, `model_id` bigint unsigned not null) default character set utf8mb4 collate 'utf8mb4_unicode_ci'  
[2020-06-28 15:10:32] production.DEBUG: alter table `model_has_roles` add index `model_has_roles_model_id_model_type_index`(`model_id`, `model_type`)  
[2020-06-28 15:10:32] production.DEBUG: alter table `model_has_roles` add constraint `model_has_roles_role_id_foreign` foreign key (`role_id`) references `roles` (`id`) on delete cascade  
[2020-06-28 15:10:32] production.DEBUG: alter table `model_has_roles` add primary key `model_has_roles_role_model_type_primary`(`role_id`, `model_id`, `model_type`)  
[2020-06-28 15:10:32] production.DEBUG: create table `role_has_permissions` (`permission_id` bigint unsigned not null, `role_id` bigint unsigned not null) default character set utf8mb4 collate 'utf8mb4_unicode_ci'  
[2020-06-28 15:10:32] production.DEBUG: alter table `role_has_permissions` add constraint `role_has_permissions_permission_id_foreign` foreign key (`permission_id`) references `permissions` (`id`) on delete cascade  
[2020-06-28 15:10:32] production.DEBUG: alter table `role_has_permissions` add constraint `role_has_permissions_role_id_foreign` foreign key (`role_id`) references `roles` (`id`) on delete cascade  
[2020-06-28 15:10:32] production.DEBUG: alter table `role_has_permissions` add primary key `role_has_permissions_permission_id_role_id_primary`(`permission_id`, `role_id`)  
[2020-06-28 15:10:32] production.DEBUG: Seeding initial roles and permissions from the migration  
`
Helpful?

etiennemarais commented:
> do a fresh install and checked out your branch before running the Attendize installer

Okay let me try doing that. I am convinced there is a exception getting swallowed and it's not reporting on it well enough. I will try and replicate it today 🙏

justynpride commented:
@etiennemarais thanks. Trust it is easy for you to find.

johannac commented:
@etiennemarais I tested this a bit and found the problem.

After running `composer install` and then `php artisan migrate` on my existing local docker set up, in the debug log I got the message 'Seeding initial roles and permissions from the migration' so I knew it had gotten to line 94 in the migration file, where it tries to call the seeder. And then it was just hanging. 

That lead me to this SO post talking about doing this same thing https://stackoverflow.com/questions/28463574/laravel-seed-after-migrating

Then in the comments there is this post https://stackoverflow.com/questions/50458656/database-gets-stuck-in-migration-with-seeder-on-production-with-force-in-larav

So I added both the changes they suggested and it works now. Line 94 in the first migration file now reads: 

`Artisan::call('db:seed', array('--class' => 'RolesAndPermissionsSeeder', '--force' => true));`

I also had to manually drop the new tables from mysql first. Since the way the migration was hanging, it left the migration in a state where the tables were there but the migration was not marked as complete, so it would try and run it again and that's how you get the table already exists error from above. This would only be a problem for people who ran the migration before we fix line 94 as above, so only me and @justynpride while testing this PR. 

```sql
drop table model_has_permissions;
drop table model_has_roles;
drop table role_has_permissions;
drop table roles; 
drop table permissions; 
```

johannac commented:
After the migration has now succeeded, when I go to the UI I get a 500 error with this in the logs: 

`[2020-06-30 21:40:17] production.ERROR: Call to a member function events() on null {"userId":2,"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalThrowableError(code: 0): Call to a member function events() on null at /usr/share/nginx/html/app/Http/Controllers/IndexController.php:35)`

On that line is `$events = $organiser->events()` which is the same as `Auth::user()->organizer->events()`. This part in the IndexController is new code introduced in this PR, so I think the organiser model is not hooked up to the user model correctly yet.

johannac commented on resources/views/errors/500.blade.php:
```diff
@@ -1,28 +1,5 @@
\ No newline at end of file
+@extends('errors::minimal')
+
+@section('title', @lang("error.back_soon"))
```
I'm also getting an error here: 

`[2020-06-30 21:50:30] production.ERROR: Call to undefined function lang() (View: /usr/share/nginx/html/resources/views/errors/500.blade.php) {"userId":2,"exception":"[object] (Facade\\Ignition\\Exceptions\\ViewException(code: 0): Call to undefined function lang() (View: /usr/share/nginx/html/resources/views/errors/500.blade.php) at /usr/share/nginx/html/resources/views/errors/500.blade.php:3)`

etiennemarais replied:
Thanks I will fix this

etiennemarais replied:
[Fix](https://github.com/Attendize/Attendize/pull/803/commits/dadfd5f21e872d13d88375cfcf421c12c8946de6)

justynpride commented:
@etiennemarais  @johannac Is it likely this might be completed and merged into the next release?

etiennemarais commented on database/migrations/2020_04_09_063112_create_permission_tables.php:
```diff
@@ -0,0 +1,117 @@
+
+        // Assign roles to all previous users in the system.
+        Log::debug("Seeding initial roles and permissions from the migration");
+        Artisan::call('db:seed', array('--class' => RolesAndPermissionsSeeder::class));
```
Thanks I will amend

etiennemarais replied:
[Fix](https://github.com/Attendize/Attendize/pull/803/commits/e578b64c4b4ab5b06f573b2b803cc64316717534)

etiennemarais commented on app/Http/Controllers/IndexController.php:
```diff
@@ -13,6 +14,47 @@ class IndexController extends Controller
+        $sort_by = (in_array($request->get('sort_by'), $allowed_sorts) ? $request->get('sort_by') : 'start_date');
+
+        // If user can manage events, then they can see all events, otherwise just their own
+        $events = $organiser->events()
```
Thanks I will add a check around the events to gracefully fall back if there are none

etiennemarais replied:
This ended up being an issue of data not linking up correctly on installation. My local test cases covered the previous users in the system and not really the fresh install. Here is the [fix for newly installed versions](https://github.com/Attendize/Attendize/pull/803/commits/96836eff03786c646667505d5de68f6c158d9cc6)

etiennemarais commented:
Sorry It's been a rough couple of weeks trying to get a massive work deadline out and the kids being sick. I will action these as soon as I can. Thank you for the review @johannac 🙏

johannac commented:
No worries at all @etiennemarais! We all understand life happens sometimes. Hope the kids are feeling better!

Commit:
Fix the error pages translations method

Commit:
Force the seed class to run on the migration to add permissions and roles tables

Commit:
Update the installation flow to include checks for first run and assign a super admin user on install as well as assign the newly created organiser to the user. This should only have to happen once though. New users will get assigned to their organiser in the manage account section fine

Commit:
Merge branch 'develop' into feature/roles-and-permissions

 Conflicts:
	VERSION
	app/Http/Controllers/ManageAccountController.php
	app/Http/Middleware/FirstRunMiddleware.php

etiennemarais commented:
@johannac I fixed those errors. Please can you test on your end? I commented about the commits that fixed each issue. 🙏

justynpride commented:
@etiennemarais Sorry to hear about your kids being unwell. Trust they are all feeling better - it has a big impact on life! 

I've just run the following. 
1. A fresh Attendize install that works fine. 
2. Checkout the roles-and-permissions branch
3. Ran composer install and then php artisan migrate - all run and complete fine.

When I go to my Attendize system I get this error displayed:

`Symfony\Component\Debug\Exception\FatalThrowableError
Call to a member function events() on null
App\Http\Controllers\IndexController::showIndex
app/Http/Controllers/IndexController.php:35
`

The logs show: 

`[2020-07-14 17:36:54] production.DEBUG: select * from `users` where `id` = ? and `users`.`deleted_at` is null limit 1 [1] 
[2020-07-14 17:36:54] production.DEBUG: select `roles`.*, `model_has_roles`.`model_id` as `pivot_model_id`, `model_has_roles`.`role_id` as `pivot_role_id`, `model_has_roles`.`model_type` as `pivot_model_type` from `roles` inner join `model_has_roles` on `roles`.`id` = `model_has_roles`.`role_id` where `model_has_roles`.`model_id` in (1) and `model_has_roles`.`model_type` = ? ["App\\Models\\User"] 
[2020-07-14 17:36:54] production.DEBUG: select `permissions`.*, `model_has_permissions`.`model_id` as `pivot_model_id`, `model_has_permissions`.`permission_id` as `pivot_permission_id`, `model_has_permissions`.`model_type` as `pivot_model_type` from `permissions` inner join `model_has_permissions` on `permissions`.`id` = `model_has_permissions`.`permission_id` where `model_has_permissions`.`model_id` in (1) and `model_has_permissions`.`model_type` = ? ["App\\Models\\User"] 
[2020-07-14 17:36:54] production.DEBUG: select * from `organisers` where ((organisers.account_id = 1))  
[2020-07-14 17:36:54] production.DEBUG: select * from `events` where `events`.`organiser_id` in (1) and `events`.`deleted_at` is null  
[2020-07-14 17:36:54] production.DEBUG: select * from `organisers` where `organisers`.`id` = ? limit 1 [0] 
[2020-07-14 17:36:54] production.ERROR: Call to a member function events() on null {"userId":1,"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalThrowableError(code: 0): Call to a member function events() on null at /home/tickets.book-in.me/public_html/Attendize/app/Http/Controllers/IndexController.php:35)
`

Don't know if this helps?

etiennemarais commented:
Did you run a attendize install? or did you have a user and database that already existed when you opened it up? Perhaps I should try and scan for previous users and assign them the first organiser id I can find. That is the issue, the users in the DB do not have the organiser relation setup via the `users.organiser_id` field in the database so it fails to do that query on the events even though there are no events to start with. 

What if you assign that user to an organiser or do a complete fresh attendize install by removing the old db locally, it should work?

justynpride commented:
@etiennemarais There was a user and database that already existed from my step 1. If I look at that user in the Db than the organiser_id is blank. If I enter 1 into that field I can login to the system. The user only appears to have access to create events, and can't add users, organisations or payment details. 

I've not done a fresh install just on this branch yet, just from master than then checking out this branch onto it.

justynpride commented:
@etiennemarais I have just done a fresh install. I don't know if this is helpful:

1. Cloned from main repo
2. Checked out the feature/roles-and-permissions branch (git pulled to triple check latest updates).
3. Followed the usual Attendize install instructions and on an empty db.

At the end of the web installer I get the error:

`Doctrine\DBAL\Schema\SchemaException
There is no column with name 'recipients' on table 'messages'.

Doctrine\DBAL\Schema\SchemaException::columnDoesNotExist
vendor/doctrine/dbal/lib/Doctrine/DBAL/Schema/SchemaException.php:85`

`[2020-07-15 00:12:12] production.DEBUG: select * from `users` where `id` = ? and `users`.`deleted_at` is null limit 1 [1] 
[2020-07-15 00:12:12] production.DEBUG: select `roles`.*, `model_has_roles`.`model_id` as `pivot_model_id`, `model_has_roles`.`role_id` as `pivot_role_id`, `model_has_roles`.`model_type` as `pivot_model_type` from `roles` inner join `model_has_roles` on `roles`.`id` = `model_has_roles`.`role_id` where `model_has_roles`.`model_id` in (1) and `model_has_roles`.`model_type` = ? ["App\\Models\\User"] 
[2020-07-15 00:12:12] production.DEBUG: select `permissions`.*, `model_has_permissions`.`model_id` as `pivot_model_id`, `model_has_permissions`.`permission_id` as `pivot_permission_id`, `model_has_permissions`.`model_type` as `pivot_model_type` from `permissions` inner join `model_has_permissions` on `permissions`.`id` = `model_has_permissions`.`permission_id` where `model_has_permissions`.`model_id` in (1) and `model_has_permissions`.`model_type` = ? ["App\\Models\\User"] 
[2020-07-15 00:12:12] production.DEBUG: select * from `organisers` where ((organisers.account_id = 1))  
[2020-07-15 00:12:12] production.DEBUG: select * from `organisers` where `organisers`.`id` = ? limit 1 [0] 
[2020-07-15 00:12:12] production.ERROR: Call to a member function events() on null {"userId":1,"exception":"[object] (Symfony\\Component\\Debug\\Exception\\FatalThrowableError(code: 0): Call to a member function events() on null at /home/tickets.book-in.me/public_html/Attendize/app/Http/Controllers/IndexController.php:35)
`

etiennemarais commented:
> I've not done a fresh install just on this branch yet, just from master than then checking out this branch onto it.

I think I can fix this by assigning `super admin` roles to all legacy users in the database. At the moment I am assigning `user` roles and they don't have permission to make new users unless you flip their role IDs in the database. Let me do that. I also need to see then if any organisers exist and assign them to the users. It's a bit of a chicken/Egg problem so I will try and sort that out today. Sorry about this, it's weird edge case 🙏

Commit:
Update the permissions seed to assign super admin roles to all legacy users. Update the organiser ID migration to look for the organiser in the same account as that belongs to the user

etiennemarais commented:
Okay so that seems to be working, here are my test steps:

1. Checkout `develop` and do a fresh composer install 
2. Make sure the `installed` file is deleted so I can run a fresh install - this will do the initial migrations and seeding
3. Sign up with my first user
4. Get redirected to the login screen and login with that user
5. Get asked to create an organiser - create one
6. See the dashboard 👍 
7. Checkout `feature/roles-and-permissions`
8. Run `composer install`
9. Run `php artisan migrate`

### Output 
```
➜ php artisan migrate
**************************************
*     Application In Production!     *
**************************************

 Do you really wish to run this command? (yes/no) [no]:
 > yes

Migrating: 2020_04_09_063112_create_permission_tables
Migrated:  2020_04_09_063112_create_permission_tables (0.32 seconds)
Migrating: 2020_04_14_193118_add_organiser_to_users
Migrated:  2020_04_14_193118_add_organiser_to_users (0.07 seconds)
```

### Log output
```
[2020-07-15 09:35:24] production.DEBUG: Seeding initial roles and permissions from the migration
[2020-07-15 09:35:25] production.DEBUG: Found 1 users to retrofit default role to
[2020-07-15 09:35:25] production.DEBUG: Assigning super admin role to legacy user: [ID:1]
[2020-07-15 09:35:25] production.DEBUG: Finished Seeding initial roles and permissions
[2020-07-15 09:35:25] production.DEBUG: Found 1 users to retrofit organiser roles to
[2020-07-15 09:35:25] production.DEBUG: Assigning organiser [ID:1] to legacy user [ID:1]
```

This will now scan for legacy users that exist based of an initial installation and assign them the correct role/organiser. Your user should now be able to add more users and assign roles in the UI. 🙏

justynpride commented:
@etiennemarais Perfect, I'm in and able to access the new feature. Thanks for your help with this. I'll start having a more thorough work through it all next week. 

A couple of things that come to mind from my initial look, which might be helpful additions or items for future consideration:
1. Delete or de-activate access for a user - people come and go in organisations after all. 
2. Invite email resend - If an invite email wasn't received or deleted by error if would be great to be able to resend the invite, as we can't delete the user and recreate at the moment.

etiennemarais commented:
@justynpride That is great news. I am glad we have now traversed through the edge cases. Thank you for the valuable input 🙏 I will have a look at those 2 enhancements, what are your thoughts @johannac ?

justynpride commented:
@etiennemarais Glad to have been of help, rather than hinderance!

justynpride commented:
@etiennemarais I’ve found that the checkin user can also add an event - ie the add event button is visible. Is this intentional, or error?

etiennemarais commented:
@justynpride no that should not be the case, I will amend thanks 🙏

c-nv-s commented:
> Okay so that seems to be working, here are my test steps:
> 
>     1. Checkout `develop` and do a fresh composer install
> 
>     2. Make sure the `installed` file is deleted so I can run a fresh install - this will do the initial migrations and seeding
> 
>     3. Sign up with my first user
> 
>     4. Get redirected to the login screen and login with that user
> 
>     5. Get asked to create an organiser - create one
> 
>     6. See the dashboard +1
> 
>     7. Checkout `feature/roles-and-permissions`
> 
>     8. Run `composer install`
> 
>     9. Run `php artisan migrate`
> 
> 
> ### Output
> 
> ```
> ➜ php artisan migrate
> **************************************
> *     Application In Production!     *
> **************************************
> 
>  Do you really wish to run this command? (yes/no) [no]:
>  > yes
> 
> Migrating: 2020_04_09_063112_create_permission_tables
> Migrated:  2020_04_09_063112_create_permission_tables (0.32 seconds)
> Migrating: 2020_04_14_193118_add_organiser_to_users
> Migrated:  2020_04_14_193118_add_organiser_to_users (0.07 seconds)
> ```
> 
> ### Log output
> 
> ```
> [2020-07-15 09:35:24] production.DEBUG: Seeding initial roles and permissions from the migration
> [2020-07-15 09:35:25] production.DEBUG: Found 1 users to retrofit default role to
> [2020-07-15 09:35:25] production.DEBUG: Assigning super admin role to legacy user: [ID:1]
> [2020-07-15 09:35:25] production.DEBUG: Finished Seeding initial roles and permissions
> [2020-07-15 09:35:25] production.DEBUG: Found 1 users to retrofit organiser roles to
> [2020-07-15 09:35:25] production.DEBUG: Assigning organiser [ID:1] to legacy user [ID:1]
> ```
> 
> This will now scan for legacy users that exist based of an initial installation and assign them the correct role/organiser. Your user should now be able to add more users and assign roles in the UI. pray

just tried to follow these steps, however the application still seems to produce an error following the migrations in the last step.
see stack trace at: https://flareapp.io/share/q5YygR5X
or perhaps I've just not followed the steps correctly

justynpride commented:
@etiennemarais This is slightly outside of the scope of this original dev, but would it be worth considering adding an additional level of user? I was thinking a booking contact, or similar role. It could be used by others down the line for allowing user registration for orders, and then logging in to view their order history for events, or for secure order changes such as for #858.

justynpride commented:
@etiennemarais I'm loving this feature. I put an issue up a while back that this relates to #800. From what I can see this feature helps towards this, but I don't think fully does. From what I can see a super admin can view all organizers, but there isn't a way to have the same abilities as the super admin, but limit the user to just one organizer. Does that make sense? 

My ideal is for a universal super admin that can access all organizers, and for there to then be a super admin that can be restricted to a selected organizer. There might be a better term for the latter option, but I use it to detail the type of access it should have. 

Currently when you create a user you can specify the organiser which then determines access. Would it be helpful for this  to determine for the Super Admin access I've mentioned? It could be listed as a dropdown in the user panel, with a 'Access All' for the main Super User.

justynpride commented:
Any progress on this being reviewed and merged?

johannac commented:
Yes, my goal is to get it into the next release going out in May. @justynpride @publicarray have you both been using this feature in production for a while now? It works fine so far from what you've seen?

justynpride commented:
> Yes, my goal is to get it into the next release going out in May. @justynpride @publicarray have you both been using this feature in production for a while now? It works fine so far from what you've seen?

@johannac That's helpful to know. Thanks for all you do on this project. I've not been using this in production yet. I've just had another more thorough review of it. From what I can see and test it works as you expect. However if you have multiple organizations using the system then this won't quite work as you might want it to. Basically you want the main organiser account to only be able to see and manage the users for their organization. I just realized that a super admin can see and edit the users for other organizations. 

A couple of helpful features for the future could be: 

1. Delete or de-activate access for a user - people come and go in organisations after all.
2. Invite email resend - If an invite email wasn't received or deleted by error if would be great to be able to resend the invite, as we can't delete the user and recreate at the moment.

quentincaffeino commented:
Also tested this feature, planning to use in staging for a while.

Besides what already was mentioned by @justynpride I found that you can't change `Can Manage Events` after user is created. (Un)Checking checkbox does nothing it seems, but I haven't dug into code of it yet.

Other than these 3 things, what else needs to be addressed? Or only thing preventing this pr from being merged is that it isn't tested enough?

quentincaffeino committed:
Merge with latest develop

justynpride commented:
@quentincaffeino personally I’d love to see this feature merged in, especially once some of these items have been addressed. Those items are to reduce end user frustration! 

One thing I’d love to see is a booking contact level so that a user can log in and make orders and review previous orders, in the same way that you can with Eventbrite and other systems. It feels like a next obvious use for this feature.

quentincaffeino commented:
Sorry, my bad, updating works. Forgot to rebuild js.

quentincaffeino committed:
Roles and Permissions: Deactivate, Restore and Delete users

quentincaffeino committed:
merge with origin/feature/roles-and-permissions

quentincaffeino committed:
Merge branch 'feature/roles-and-permissions' into feature/roles-and-permissions-deleting-users

quentincaffeino committed:
Added prebuild backend js

quentincaffeino committed:
Created dropdownmenu to make things look better and to allow to fit more buttons

quentincaffeino committed:
Added `Send invitation` button

quentincaffeino committed:
fix: if user has no roles whole settings is broken, check if user has roles

quentincaffeino committed:
Merge branch 'develop' into feature/roles-and-permissions-deleting-users

quentincaffeino committed:
fix: password is empty on user save

johannac committed:
Merge pull request #957 from quentincaffeino/feature/roles-and-permissions-deleting-users

Roles and Permissions: Deactivate, Restore, Delete and send invitations to users

johannac committed:
Merge branch 'develop' into feature/roles-and-permissions

johannac commented:
@quentincaffeino @justynpride just checking -- apart from general QA and testing, there is only 1 outstanding functionality issue with this system which was originally noted here (https://github.com/Attendize/Attendize/pull/957#issuecomment-895236900) which is what to do about the Super Admin role and determining how it should work across repositories -- is that right?

justynpride commented:
> @quentincaffeino @justynpride just checking -- apart from general QA and testing, there is only 1 outstanding functionality issue with this system which was originally noted here ([#957 (comment)](https://github.com/Attendize/Attendize/pull/957#issuecomment-895236900)) which is what to do about the Super Admin role and determining how it should work across repositories -- is that right?

From my perspective that is correct. Thanks for looking at this.

quentincaffeino commented:
Tbh don't remember because I haven't touched Attendize since 2021

JSkye3293 commented:
Did this get pushed to the main stream? Im not seeing these changes in the current build that I have. Is there a way I need to enable them? Thanks!
